tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+           from remote op names to a dictionary from tensor names to tensors.
+           Here, element[remote_op_name] contains graph inputs.
+    remote_op_name: The remote op name of the current graph.
+    remote_op_name_to_graph_name:
+      A mapping from remote op names to graph names.
+    graph_name_to_specs:
+      A mapping from graph names to a list of ExecutionSpecs, where the order
+      of the list represents the order of execution.
+
+    # We don't have this information since it was stored in PyFunc's function.
+    graph_to_remote_op_input_name_mapping:
+      A mapping from graph names to remote op names to remote graph placeholder
+      names to parent graph input names.
+      {graph name: {remote op name: {placeholder name: input name}}}.
+
+  Returns:
+    A PCollection of results of this graph. Each element is a dictionary from
+    remote op names to a dictionary from tensor names to tensors. Here,
+    element[remote_op_name] stores graph inputs, intermediate results, and
+    graph outputs.
+  """
+  specs = graph_name_to_specs[remote_op_name_to_graph_name[remote_op_name]]
+
+  for spec in specs:
+    # Construct Beam subgraph for a subgraph layer
+    if not spec.is_remote_op:
+      step_name = ("SubgraphLayerDoFn[Graph_%s][Outputs_%s]" %
+                   (remote_op_name, '_'.join(spec.output_names)))
+      pcoll = pcoll | step_name >> beam.ParDo(
+          _SubgraphLayerDoFn(), spec, remote_op_name)
+
+    # Construct Beam subgraph for a remote op
+    else:
+      # ExecutionSpec stores one remote op.
+      child_remote_op_name = list(spec.output_names)[0]
+      step_descriptor = ("[Parent_%s][Child_%s]" %
+                         (remote_op_name, child_remote_op_name))
+
+      step_name = "LoadRemoteGraphInputs%s" % step_descriptor
+      pcoll = pcoll | step_name >> _LoadRemoteGraphInputs(
+          remote_op_name,
+          child_remote_op_name,
+          remote_op_name_to_graph_name,
+          graph_to_remote_op_input_name_mapping)
+
+      # A good place to add beam.Reshuffle()
+      step_name = "ExecuteGraph%s" % step_descriptor
+      pcoll = pcoll | step_name >> ExecuteGraph(
+          child_remote_op_name,
+          remote_op_name_to_graph_name,
+          graph_name_to_specs,
+          graph_to_remote_op_input_name_mapping)
+
+      step_name = "ExtractRemoteGraphOutput%s" % step_descriptor
+      pcoll = pcoll | step_name >> _ExtractRemoteGraphOutput(
+          remote_op_name,
+          child_remote_op_name,
+          remote_op_name_to_graph_name,
+          graph_name_to_specs)
+
+  return pcoll
+
+
+class _SubgraphLayerDoFn(beam.DoFn):
+  """DoFn that executes one subgraph layer."""
+
+  def process(
+      self,
+      # Not using mapping here because it doesn't support item assignment.
+      element: Dict[Text, Dict[Text, tf.Tensor]],
+      spec: execution_spec.ExecutionSpec,
+      remote_op_name: Text
+  ) -> Generator[Dict[Text, Dict[Text, tf.Tensor]], None, None]:
```

Comment:
Since `SendType` and `ReturnType` are `None`, the type can be shortened to: `Iterator[Dict[Text, Dict[Text, tf.Tensor]]]`.

Reply:
Addressed.